### PR TITLE
fix(distillation): reverse-KL server path NaN on variable completion length

### PR DESCRIPTION
**Pull request by k1064190**
# What does this PR do?

Fixes a NaN-gradient bug in `DistillationTrainer`'s server-backed reverse-KL loss when batches produce per-sample padding within the rectangular teacher logprob tensor.

**Trigger**: `use_teacher_server=True` + `beta == 1.0` (reverse KL) + any config where `_get_teacher_token_logprobs_from_server` emits `-inf` padding. This happens whenever per-sample `completion_offsets` differ (verified empirically with `bs=1, ga=2` and `bs=2, ga=1`). Forward loss is finite (clamped by `nan_to_num`); `grad_norm=nan` on the first optim step.

**Root cause**: `_get_teacher_token_logprobs_from_server` pads the rectangular `(B, T)` teacher tensor with `-inf` — both for shorter samples in cross-sample ragged batches *and* at per-sample head offsets (`completion_start - aligned_prompt_length`). The forward-KL server path (`_compute_server_forward_kl_loss`) tolerates this: its `p·(log_t - log_s)` form has `teacher_probs=0` at padding, so `0·(-inf)=NaN` in forward is `nan_to_num`'d to 0, and backward partials `∂/∂safe_teacher = teacher_probs = 0` zero the gradient. The JSD mixture path is safe for the same reason — `log((1-β)·student_probs + β·teacher_probs)` stays finite when `teacher_probs=0`.

Reverse-KL (`student_probs·(log_s - log_t)`) has no such protection: the multiplier `student_probs` is nonzero at padding, so `finite·(+inf)=+inf` survives into backward, and `grad_norm` becomes NaN via `sqrt(sum(inf²))`/clipping.

**Fix**: In `_compute_server_sparse_top_1_divergence_loss`, after the existing `isfinite` validation, neutralise the `-inf` sentinel at known padding positions (`labels == -100`) with a finite zero via `torch.where`, before the shared divergence helper runs. The label mask in `_reduce_divergence_loss` continues to exclude these positions from the final loss — the new block only prevents the sentinel from reaching the autograd graph.

This is the minimal surgical fix: the shared helper `_compute_sparse_top_1_divergence_loss` is untouched, the `-inf` sentinel contract (used by the existing `isfinite` validator to flag missing-required-data) is preserved, and forward-KL / JSD paths are not affected.

**Tests** (`tests/experimental/test_distillation_trainer.py`, new file — trainer had no dedicated tests):
- Sentinel contract at the server getter.
- Mask pattern in isolation: `_add_tail_bucket` + `_jsd_divergence(beta=1)` post-mask produces finite forward AND backward.
- End-to-end `DistillationTrainer.train()` with mocked `VLLMClient`, parametrized over `bs=1, ga=2` and `bs=2, ga=1`. Both fail (`grad_norm=nan`) without the fix; both pass with it.

`pytest tests/experimental/test_distillation_trainer.py -v`: 4/4 pass.

**Env** (`trl env`):
```
- Platform: Linux-5.14.0-427.22.1.el9_4.x86_64-x86_64-with-glibc2.35
- Python version: 3.11.15
- TRL version: 1.3.0.dev0+3c0d9ae
- PyTorch version: 2.10.0+cu130
- accelerator(s): NVIDIA RTX PRO 6000 Blackwell Server Edition x3
- Transformers version: 4.57.3
- Accelerate version: 1.13.0
- Datasets version: 4.8.4
- HF Hub version: 0.36.2
- bitsandbytes version: 0.49.2
- DeepSpeed version: 0.18.9
- Liger-Kernel version: 0.7.0
- PEFT version: 0.19.1
- vLLM version: 0.17.1
```

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

## AI writing disclosure

- [ ] No AI usage: the PR was written entirely by a human.
- [x] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.
- [ ] AI-generated: the PR was mostly or fully generated by an AI tool.

---

> [!NOTE]
> **Medium Risk**
> Touches core distillation loss math in the teacher-server path; while the change is targeted to padding positions, it can affect training stability/metrics for reverse-KL setups if masking semantics are wrong.
> 
> **Overview**
> Fixes a NaN-gradient issue in `DistillationTrainer` when using `use_teacher_server=True` with reverse-KL (`beta=1`) on variable-length completion batches.
> 
> The server sparse top-1 loss path now **neutralizes padded `-inf` teacher logprobs** (where `labels == -100`) by replacing them with `0` before divergence computation, preventing `+inf` from entering autograd while preserving the existing “missing required logprobs” validation.
> 
> Adds a new `tests/experimental/test_distillation_trainer.py` suite covering: the `-inf` padding sentinel contract in `_get_teacher_token_logprobs_from_server`, finite forward/backward behavior when masking padded positions, and an end-to-end mocked teacher-server training step asserting finite `loss`/`grad_norm` for ragged batches.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit f85334af9ee965b8f93b6449e9de03c13bb80309. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: tests/experimental/test_distillation_trainer.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+    def _run_one_step(self, beta, monkeypatch):
+        from trl.generation import vllm_client as vllm_client_module
+
+        fake_client = MagicMock()
+        fake_client.get_sequence_logprobs.side_effect = _canned_teacher_logprobs
+        monkeypatch.setattr(vllm_client_module, "VLLMClient", lambda *a, **kw: fake_client)
+
+        config = DistillationConfig(
+            output_dir=self.tmp_dir,
+            per_device_train_batch_size=1,
```
[the remark this answers is not in the document]
Comment (k1064190) [reply tests/experimental/test_distillation_trainer.py]: Thanks for the review. We verified empirically and the conclusion is nuanced:

**Bugbot's observation about `bs=1` is partially correct but doesn't fully generalize.** Cross-sample ragged padding is *one* source of `-inf` in the teacher tensor, but not the only one. `_get_teacher_token_logprobs_from_server` also produces `-inf` via per-sample `completion_offsets = completion_start - aligned_prompt_length` (L1276–1283, L1299–1303): if a sample's completion starts later than the batch-wide `aligned_prompt_length`, the `arr[i, 0:offset]` range is left as `-inf`. This triggers with `bs=1` too.

Confirmed by 2×2 experiments (fix retained vs. fix removed × `bs=1, ga=2` vs. `bs=2, ga=1`):

| config | fix retained | fix removed |
|---|---|---|
| `bs=1, ga=2`, beta=1.0 | PASS | **FAIL `grad_norm=nan`** |
| `bs=2, ga=1`, beta=1.0 | PASS | **FAIL `grad_norm=nan`** |
| `bs=1, ga=2`, beta=0.5 | PASS | PASS (vacuous) |
| `bs=2, ga=1`, beta=0.5 | PASS | PASS (vacuous) |

So the end-to-end test *was* a valid regression guard in both configs — but to strengthen coverage we've parametrized it over both (5d3d085) and dropped the JSD (beta=0.5) end-to-end test entirely because it's vacuous in every config.

**Why JSD is vacuous**: in `_jsd_divergence` the mixture branch routes student gradients through `log((1-β)·student_probs + β·teacher_probs)`, which stays finite when `teacher_probs=0` at padding. The `-inf` only appears in the `kl_teacher` term (multiplied by `teacher_probs=0`, giving `0·(-inf)=NaN` forward, zero-gradient backward via `∂/∂safe_teacher = teacher_probs = 0`). Reverse-KL has no such routing — its `student_probs·(log_s - log_t)` multiplies `-inf` by a nonzero `student_probs`, preserving `+inf` into backward. This is why only reverse-KL needs the fix and only the reverse-KL test is kept as a regression guard.

The PR body has been updated to reflect the corrected trigger condition.

File: trl/experimental/distillation/distillation_trainer.py [hunk 2]
```diff
@@ -1354,6 +1358,22 @@ def _compute_server_sparse_top_1_divergence_loss(
...
+        # Padding positions (labels == -100) within the batch's completion_length carry the
+        # -inf sentinel assigned by _get_teacher_token_logprobs_from_server for shorter samples
+        # in variable-length batches. The label mask in _reduce_divergence_loss already
+        # excludes these positions from the final loss, but their -inf values still propagate
+        # through _add_tail_bucket (producing teacher distributions [-inf, 0]) and
+        # _jsd_divergence (producing +inf in forward, clamped by nan_to_num, but NaN in
+        # backward because autograd's chain rule does not respect nan_to_num). Neutralise the
+        # sentinel at known padding positions before the shared divergence helper runs,
+        # mirroring the masking applied by _compute_server_forward_kl_loss for the forward-KL
+        # path.
```
Comment (cmpatino) [inline trl/experimental/distillation/distillation_trainer.py]: Can you please reduce the verbosity of this comment? A line or two describing what the masks do should be enough

File: trl/experimental/distillation/distillation_trainer.py [hunk 3]
```diff
@@ -1354,6 +1358,22 @@ def _compute_server_sparse_top_1_divergence_loss(
...
+        pad_mask_2d = ~required
+        pad_mask_3d = pad_mask_2d.unsqueeze(-1)
+        zero = torch.zeros((), dtype=topk_teacher_lps.dtype, device=topk_teacher_lps.device)
+        topk_teacher_lps = torch.where(pad_mask_3d, zero, topk_teacher_lps)
+        actual_teacher_lps = torch.where(pad_mask_2d, zero, actual_teacher_lps)
```
Comment (cmpatino) [inline trl/experimental/distillation/distillation_trainer.py]: Can we do something like the snippet below to avoid defining the `zero` tensor?

```python
topk_teacher_lps = torch.where(pad_mask_3d, 0.0, topk_teacher_lps)
actual_teacher_lps = torch.where(pad_mask_2d, 0.0, actual_teacher_lps)
```

**Review by cmpatino — CHANGES_REQUESTED** [review]
Thank you for the contribution! This is a good check to prevent NaNs while training.

I left some comments before approving the merge of the PR.

File: tests/experimental/test_distillation_trainer.py [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+            model=model,
+            args=config,
+            train_dataset=_variable_length_dataset(),
+            processing_class=self.tokenizer,
+        )
+        trainer.teacher_client = fake_client
+        trainer.train()
+        return [rec for rec in trainer.state.log_history if "grad_norm" in rec]
+
+    @pytest.mark.parametrize(("bs", "ga"), [(1, 2), (2, 1)])
```
Comment (cmpatino) [inline tests/experimental/test_distillation_trainer.py:177]: Should we add `@pytest.mark.slow` to this test? It downloads a model and runs `trainer.train()` for one step, which could be expensive to run every time.

File: trl/experimental/distillation/distillation_trainer.py (hunk 2 above)
Comment (k1064190) [reply trl/experimental/distillation/distillation_trainer.py]: Done. trimmed to two lines: what the mask does (replace -inf at padding with 0) and why (prevents reverse-KL from leaking +inf into backward).

File: tests/experimental/test_distillation_trainer.py (hunk 4 above)
Comment (k1064190) [reply tests/experimental/test_distillation_trainer.py:177]: Added. Only the end-to-end test (model download + trainer.train() step) is marked slow. the two lower-level tests (sentinel contract, isolated mask + divergence finite check) remain in the default suite.

File: trl/experimental/distillation/distillation_trainer.py (hunk 3 above)
Comment (k1064190) [reply trl/experimental/distillation/distillation_trainer.py]: Dropped the zero scalar tensor. torch.where broadcasts the 0.0 literal to the tensor's dtype/device automatically.

**Review by cmpatino — APPROVED** [review]

**Comment by cmpatino** [comment]
Changes LGTM! Feel free to merge the PR
